config.txt


# PARTICLES #

# How many particles to spawn, proportionate to the exploder's mass.
# It the game runs slowly after a collision, this is the variable to turn down.
particle-ratio = 80

# How large should the particles be? 1 is default.
particle-size = 1.0

# How should particles behave? 
# simple: particles will move in straight lines until off screen.
#         Use this option if you encounter performance problems.
# orbit:  particles will orbit attractors.
# gravity-field: Particles will display showing their gravitational pulls.
#particle-behaviour = simple
particle-behaviour = orbit
#particle-behaviour = gravity-field

# What percentage should particles be visible?
particle-opacity = 30

# PATH PREDICTION #

# A prediction line, showing the path enemies will fallow, can be drawn to help
# beginners learn how to play and manipulate enemies.

# How time increments to predict. To turn the prediction line off, set this to
# 0.  Setting it too high may cause slow down or crash, but it can look pretty
# interesting if you set it to different values. Try 1000, 50, etc.
prediction-length = 0

# How precise should the predictions be?
# Lower value equals higher precision and a shorter, but smoother prediction
# line. 
prediction-precision = 10

# OTHER OPTIONS #

# A line can be drawn from the enemy to the player to help you understand the
# relationship of all the objects on screen.
gravity-line = 0

# These options sets whether or not to draw an arrow representing the
# velocities or accelerations of your enemies. 0 for off, non-zero for on.
velocity-arrow     = 0
acceleration-arrow = 0

# Add motion blur; 0 for off, non-zero for on. 
# This is mostly eye candy and can make it harder to play.
motion-blur = 0

# How many high scores should be saved?
nHighScores = 10

# Display frames per second? (0=off, else on.)
fps = 0

# Music and Sound volumes: Maximum is 128, minimum is zero.
music-volume = 80
sfx-volume   = 80


config-default.txt
# How many particles to spawn, proportionate to the exploder's mass.
# It the game runs slowly after a collision, this is the variable to turn down.
particleRatio = 100

# A prediction line, showing the path enemies will fallow assuming the player
# stands still, can be drawn to help beginners learn how to play and manipulate
# enemies. This option is recommended as off because it creates visual
# distractions making you less likely to notice an enemy about to hit you. Keep
# it on until you feel comfortable without it.

# How far into the future to predict. To turn the prediction line off, set this
# to 0.  Setting it high may cause slow down or crash, but it can look pretty
# interesting if you set it high, like 1000.
predictionLength = 0

# Lower value equals higher precision and a shorter, but smoother prediction
# line. In physics terms: this game runs on a "fixed-time-step" meaning that
# the physics is integrated with a constant change in time. Instead of
# predicting every singe state the object could be in for each time-step, i
# skip an arbitrary number of them. This is the number of skipped states.
predictionPrecision = 10

# A line can be drawn from the enemy to the player to help you understand when
# direction the enemies are accelerating. This is recommended for intermediate
# player who do not need the prediction line.
gravityLine = 0

# This option sets whether or not to draw an arrow representing the
# velocities of accelerations of your enemies. 0 for off, non-zero for on.
velocityArrow     = 0
accelerationArrow = 0

# Add motion blur; 0 for off, non-zero for on. This is mostly eye candy and can
# make it harder to play.
motionBlur = 0

# FOR TESTING PURPOSES ONLY, you may set the scale of the game. This changes
# the size of everything on screen except text and the size of the screen.
# Set to 1 for normal size, 0.5 for half size, 2 for double, etc.
scale = 0.6

# How many high scores should be saved.
nHighScores = 10

# Display frames per second? (0=off, else on.)
fps = 1

# How should particles behave?
particle-behaviour = orbit
#particle-behaviour = gravity-field

# Music volume: Maximum is 128, minimum is zero.
music-volume = 100


Config.cpp

#include "Config.h"
#include "Parsing.h"

#include <fstream>
#include <algorithm>

void Config::set_defaults()
{
    vars["particle-ratio"]     = "200";
    vars["particle-size"]      = "1";
    vars["particle-opacity"]   = "30";
    vars["particle-behaviour"] = "orbit";

    vars["prediction-length"]    = "0";
    vars["prediction-precision"] = "10";

    vars["gravity-line"]       = "0";
    vars["velocity-arrow"]     = "0";
	vars["acceleration-arrow"] = "0";

    vars["motion-blur"] = "0";

    vars["nHighScores"] = "10";

    vars["fps"] = "0";

    // Values out of 128 for SDL_mixer.
    vars["music-volume"] = "100";
    vars["sfx-volume"]   = "128"; 
}

Config::Config()
{
    set_defaults();
}

Config::Config( const std::string& filename )
{
    reload( filename );
}

bool Config::reload( const std::string& filename )
{
    // To ensure nothing is left uninitialized.
    set_defaults();

    std::ifstream cfg( filename.c_str() );

    std::string line;
    while( std::getline( cfg, line ) ) 
    {
        rm_comments(   &line );
        rm_whitespace( &line );

        if( line.size() <= 1 )
            continue;

        // Check for errors in var after checking it against the possible
        // choices. That way, bad expression and unknown handle errors 
        // can be handled together.
        Variable var = evaluate_expression( line );

        if( vars.find(var.handle) != vars.end() ) {
            vars[ var.handle ] = var.value;
        } else {
            static std::ofstream out( "config.error" );
            out << "Unknown valName (" << var.handle << "}\n";
            out << "Line: \"" << line << "\"\n\n";
        }
    }

    return cfg;
}

std::string& Config::operator [] ( const std::string& handle )
{
    return vars[ handle ];
}

const std::string& Config::operator [] ( const std::string& handle ) const
{
    static const std::string ZILCH = "";
    Vars::const_iterator it = vars.find( handle );
    if( it != vars.end() )
        return it->second;
    else
        return ZILCH;
}
